// babel-plugin-flow-react-proptypes/lib/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PLUGIN_NAME = exports.$debug = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.isExact = isExact;
exports.makeLiteral = makeLiteral;
exports.getExportNameForType = getExportNameForType;
exports.containsReactElement = containsReactElement;
exports.hasReactElementTypeAnnotationReturn = hasReactElementTypeAnnotationReturn;

var _babelTypes = require('babel-types');

var t = _interopRequireWildcard(_babelTypes);

var _babelTraverse = require('babel-traverse');

var _babelTraverse2 = _interopRequireDefault(_babelTraverse);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var $debug = exports.$debug = function $debug() {};
// export const $debug = console.error.bind(console);

var PLUGIN_NAME = exports.PLUGIN_NAME = 'babel-plugin-flow-react-proptypes';

function isExact(node) {
  return node.id.name === '$Exact';
}

function makeLiteral(value) {
  if (typeof value === 'string') return t.stringLiteral(value);else if (typeof value === 'number') return t.numericLiteral(value);else if (typeof value === 'boolean') return t.booleanLiteral(value);else {
    $debug('Encountered invalid literal', value);
    throw new TypeError('Invalid type supplied, this is a bug in ' + PLUGIN_NAME + ', typeof is ' + (typeof value === 'undefined' ? 'undefined' : _typeof(value)) + ' with value ' + value);
  }
}

function getExportNameForType(name) {
  return 'bpfrpt_proptype_' + name;
}

function containsReactElement(node) {
  var fakeRoot = {
    type: 'File', program: {
      type: 'Program',
      sourceType: 'module',
      body: [node]
    }
  };
  var matched = false;

  (0, _babelTraverse2.default)(fakeRoot, {
    JSXElement: function JSXElement(path) {
      matched = true;
      path.stop();
    },
    CallExpression: function CallExpression(path) {
      if (matched) {
        path.stop();
        return;
      }

      var node = path.node;
      var callee = node.callee;

      if (callee.type !== 'MemberExpression') return;
      if (callee.object && callee.object.name === 'React' && callee.property && callee.property.name === 'createElement') {
        matched = true;
        path.stop();
      }
    }
  });
  return matched;
}

function hasReactElementTypeAnnotationReturn(node) {
  if (node.type !== 'ArrowFunctionExpression') {
    return false;
  }
  if (!node.returnType || node.returnType.type !== 'TypeAnnotation') {
    return false;
  }

  var type = node.returnType.typeAnnotation;
  if (type.type === 'GenericTypeAnnotation') {
    if (type.id && type.id.name === 'ReactElement') {
      return true;
    } else {
      return false;
    }
  }

  return false;
}
